datafusion/core/src/execution/context/mod.rs

```diff
@@ -1261,7 +1313,38 @@ impl QueryPlanner for DefaultQueryPlanner {
             .await
     }
 }
+/// A pluggable interface to handle `CREATE FUNCTION` statements
+/// and interact with [SessionState] to registers new udf, udaf or udwf.
+
+#[async_trait]
+pub trait FunctionFactory: Sync + Send {
```

Comment:
The last thing standing is trait name, I'm not sure if `FunctionFactory` is the best name for it. It would be appropriate if this PR is part of spark framework :) 

 any suggestions @alamb ?

Reply:
I think the name is good and describes what this trait does -- namely it creates functions. I can't think of anything better or more appropriate